source/common/http/conn_pool_grid.cc

```diff
@@ -71,18 +80,69 @@ void ConnectivityGrid::WrapperCallbacks::ConnectionAttemptCallbacks::onPoolFailu
   parent_.onConnectionAttemptFailed(this, reason, transport_failure_reason, host);
 }
 
+void ConnectivityGrid::WrapperCallbacks::attemptHttp3HappyEyeballs() {
+  // Swap the v4/v6 pool for a v6/v4 pool.
+  if (!grid_.http3_alternate_pool_) {
+    grid_.http3_alternate_pool_ = grid_.createHttp3Pool(true);
+    grid_.setupPool(*grid_.http3_alternate_pool_);
+    grid_.pools_.push_back(grid_.http3_alternate_pool_.get());
+  }
+  has_tried_http3_alternate_address_ = true;
+
+  auto attempt = std::make_unique<ConnectionAttemptCallbacks>(*this, *grid_.http3_alternate_pool_);
+  LinkedList::moveIntoList(std::move(attempt), connection_attempts_);
+  // kick off a new stream attempt.
+  connection_attempts_.front()->newStream();
+}
+
+bool ConnectivityGrid::WrapperCallbacks::shouldAttemptHttp3HappyEyeballs() {
+  // Don't connect if the grid is shutting down.
+  if (grid_.destroying_) {
+    return false;
+  }
+  // Make sure each wrapper callbacks only tries HTTP/3 alternate addresses
+  // once.
+  if (has_tried_http3_alternate_address_) {
+    return false;
+  }
+  // Branch on reloadable flags.
+  if (!Runtime::runtimeFeatureEnabled("envoy.reloadable_features.http3_happy_eyeballs")) {
+    return false;
+  }
+  // QUIC "happy eyeballs" currently only handles one v4 and one v6 address. If
+  // there's not multiple families don't bother.
+  if (!hasBothAddressFamilies(grid_.host_)) {
+    return false;
+  }
+  return true;
+}
+
 void ConnectivityGrid::WrapperCallbacks::onConnectionAttemptFailed(
     ConnectionAttemptCallbacks* attempt, ConnectionPool::PoolFailureReason reason,
     absl::string_view transport_failure_reason, Upstream::HostDescriptionConstSharedPtr host) {
   ENVOY_LOG(trace, "{} pool failed to create connection to host '{}'.",
             describePool(attempt->pool()), host->hostname());
+  grid_.dispatcher_.deferredDelete(attempt->removeFromList(connection_attempts_));
+
   if (grid_.isPoolHttp3(attempt->pool())) {
-    http3_attempt_failed_ = true;
+    if (shouldAttemptHttp3HappyEyeballs()) {
+      // Hot swap the original HTTP/3 connection pool for one of the alternate family.
+      attemptHttp3HappyEyeballs();
+      // Return - the above attempt will handle everything else.
+      return;
+    }
+    if (has_tried_http3_alternate_address_) {
```

Comment:
I don't think so.  If we try both we have to make sure if there's an attempt left it's only TCP as the H3 connections can fail out of order if we kick off the H3 alternate attempt due to timeout rather than expicit H3 failure
That said we can do the more complicated check each time